### PR TITLE
kie-issues#1162: Add DMN Boxed Filter expression tests

### LINKED ISSUE #1162: Add end-to-end tests for Filter Expression on the Boxed Expression Component
- [x] Create a new Story for the Filter Expression adding examples and documentation.
- [x] Create end-to-end tests for the Filter Expression;

### PR DESCRIPTION
**Pull request by jomarko**
Closes apache/incubator-kie-issues#1162

### CODE REVIEW

**Review by porcelli — APPROVED** [review]

File: packages/boxed-expression-component/tests/e2e/boxedExpressions/boxedExpressionHeader.spec.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@ import { test, expect } from "../__fixtures__/base";
 test.describe("Boxed expression header", () => {
   test.beforeEach(async ({ boxedExpressionEditor, browserName }) => {
     test.skip(
-      browserName !== "chromium",
+      browserName === "webkit",
```
Comment (ljmotta) [inline packages/boxed-expression-component/tests/e2e/boxedExpressions/boxedExpressionHeader.spec.ts]: 👍

File: packages/boxed-expression-component/tests/e2e/boxedExpressions/filter/populate.spec.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import _ from "lodash";
```
Comment (ljmotta) [inline packages/boxed-expression-component/tests/e2e/boxedExpressions/filter/populate.spec.ts]: I couldn't find the usage

File: packages/boxed-expression-component/tests/e2e/boxedExpressions/filter/populate.spec.ts
Comment (ljmotta) [inline packages/boxed-expression-component/tests/e2e/boxedExpressions/filter/populate.spec.ts:1]: We left the `populate.spec.ts` to add a complete case. On this file you could left the creation of `Rebooked Flights` example, or any other example, like the nested list one.

Add a `filterExpression.spec.ts` to check if it render correctly, you could move the create a nested filter to this file too and the respective reset/reset nested. Move the `rename` and `change data type` too.

File: packages/boxed-expression-component/stories/boxedExpressions/Filter/Filter.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+          __$$element: "literalExpression",
+          "@_id": generateUuid(),
+        },
+      },
+      match: {
+        "@_id": generateUuid(),
+        expression: {
+          __$$element: "literalExpression",
+          "@_id": generateUuid(),
+        },
```
Comment (ljmotta) [inline packages/boxed-expression-component/stories/boxedExpressions/Filter/Filter.stories.tsx]: The `base` example should be without expressions.

File: packages/boxed-expression-component/tests/e2e/boxedExpressions/filter/populate.spec.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    page,
+    stories,
+  }) => {
+    await stories.openBoxedFilter("base");
+    await page.getByText("=").first().click({ button: "right" });
+    await page.getByRole("menuitem").getByText("Reset").click();
+    await boxedExpressionEditor.selectBoxedList(page.getByText("Select expression").first());
+
+    await expect(boxedExpressionEditor.getContainer()).toHaveScreenshot("boxed-filter-nested-boxed-list.png");
+  });
```
Comment (ljmotta) [inline packages/boxed-expression-component/tests/e2e/boxedExpressions/filter/populate.spec.ts]: Please, add a more complete example here. The list choice looks a little randomic, as the filter supports any other expression as well.

File: packages/boxed-expression-component/tests/e2e/boxedExpressions/filter/populate.spec.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    await context.grantPermissions(["clipboard-read", "clipboard-write"]);
+
+    await stories.openBoxedFilter("rebooked-flights");
+    await boxedExpressionEditor.copyFilter();
+    await boxedExpressionEditor.resetFilter();
+    await boxedExpressionEditor.selectBoxedContext(page.getByText("Select expression").first());
+    await boxedExpressionEditor.pasteToUndefinedCell(page.getByText("Select expression").first());
+
+    await expect(boxedExpressionEditor.getContainer()).toHaveScreenshot("boxed-filter-copied-and-pasted-as-nested.png");
+  });
```
Comment (ljmotta) [inline packages/boxed-expression-component/tests/e2e/boxedExpressions/filter/populate.spec.ts]: Great addition. Even though we have some generic tests for the `copy/paste` on the `selection.spec.ts` I guess the suite is lacking more specific tests. Please move this to the `selection.spec.ts` inside a `Filter Expression` description block.

File: packages/boxed-expression-component/tests/e2e/__fixtures__/boxedExpression.ts [hunk 6]
```diff
@@ -89,6 +94,36 @@ export class BoxedExpressionEditor {
...
+
+  public async copyFilter(from: Page | Locator = this.page) {
+    await from.getByRole("button").getByText("Filter").click();
+    await this.page.getByRole("menuitem").getByText("Copy").click();
+  }
+
+  public async resetFilter(from: Page | Locator = this.page) {
+    await from.getByRole("button").getByText("Filter").click();
+    await this.page.getByRole("menuitem").getByText("Reset").click();
+  }
```
Comment (ljmotta) [inline packages/boxed-expression-component/tests/e2e/__fixtures__/boxedExpression.ts]: Currently, we don't have specific fixtures, which we need to revisit on the near future. For this case, I guess we could stick with what we have, and use inline `testId` we already use on some other tests.

```
await page.getByTestId("logic-type-button-test-id").click();
await page.getByRole("menuitem", { name: "copy" }).click();

or 

await page.getByRole("menuitem", { name: "reset" }).click();
```

Or we could start creating the specific fixtures. WDYT?

**Review by ljmotta — COMMENTED** [review]
@jomarko Thanks for this PR. I've made some comments inline. Also I've relized it's missing a test on the `resize.spec.ts` file. Please add one `decription` block there, we need to check the resize there.

File: packages/boxed-expression-component/stories/boxedExpressions/Filter/Filter.stories.tsx (hunk 3 above)
Comment (jomarko) [reply packages/boxed-expression-component/stories/boxedExpressions/Filter/Filter.stories.tsx]: I can not remove them
Screenshot 2024-06-21 150240

**Review by danielzhe — APPROVED** [review]
Everything in this PR is going to change to an easier way to write and read tests (except the images), but the code you made is going to be reutilized inside the API. So it is better to merge now so I can reuse this code inside the API.

File: packages/boxed-expression-component/src/expressions/FilterExpression/FilterExpressionCollectionCell.tsx [hunk 7]
```diff
@@ -52,7 +52,7 @@ export function FilterExpressionCollectionCell({
   );
 
   return (
-    <div className="filter-expression">
+    <div className="filter-expression" data-testid={"kie-tools--boxed-expression-component---filter-collection-in"}>
```
Comment (ljmotta) [inline packages/boxed-expression-component/src/expressions/FilterExpression/FilterExpressionCollectionCell.tsx]:
```suggestion
    <div className="filter-expression" data-testid={"kie-tools--boxed-expression-component--filter-collection-in"}>
```

File: packages/boxed-expression-component/src/expressions/FilterExpression/FilterExpressionMatchCell.tsx [hunk 8]
```diff
@@ -101,7 +101,7 @@ export function FilterExpressionMatchCell({
       >
         <div className={"bracket-sign"}>[</div>
       </div>
-      <div>
+      <div data-testid={"kie-tools--boxed-expression-component---filter-collection-match"}>
```
Comment (ljmotta) [inline packages/boxed-expression-component/src/expressions/FilterExpression/FilterExpressionMatchCell.tsx]:
```suggestion
      <div data-testid={"kie-tools--boxed-expression-component--filter-collection-match"}>
```

File: packages/boxed-expression-component/stories/boxedExpressions/Filter/Filter.stories.tsx (hunk 3 above)
Comment (ljmotta) [reply packages/boxed-expression-component/stories/boxedExpressions/Filter/Filter.stories.tsx]: For this case, you can use `expression: undefined!`, which will force an empty expression. With it, you can add a comment ` // SPEC DISCREPANCY: Starting without an expression gives users the ability to select the expression type.`.

File: packages/boxed-expression-component/stories/getDefaultBoxedExpressionForStories.ts [hunk 9]
```diff
@@ -235,6 +237,28 @@ export function getDefaultBoxedExpressionForStories({
...
+          __$$element: "literalExpression",
+        },
+      },
+      match: {
+        "@_id": generateUuid(),
+        expression: {
+          "@_id": generateUuid(),
+          __$$element: "literalExpression",
+        },
+      },
```
Comment (ljmotta) [inline packages/boxed-expression-component/stories/getDefaultBoxedExpressionForStories.ts]: The same here. Follow the `context` expression case, and add a `expression: undefined!` with the `SPEC DISCREPANCY` comment please.

File: packages/boxed-expression-component/tests-e2e/__fixtures__/boxedExpression.ts [hunk 10]
```diff
@@ -33,6 +33,11 @@ export class BoxedExpressionEditor {
     await from.getByText("Select expression").click();
   }
 
+  public async pasteToUndefinedCell(from: Page | Locator = this.page) {
+    this.select(from);
+    await this.page.getByRole("menuitem").getByText("Paste").click();
+  }
```
Comment (ljmotta) [inline packages/boxed-expression-component/tests-e2e/__fixtures__/boxedExpression.ts]: Hmm, shouldn't it be `pasteToSelectExpression`? I'm not sure if `undefined cell` is a good name, as we don't mention this on any other place. Maybe `pasteToEmptyExpression`? 😅

File: packages/boxed-expression-component/tests-e2e/__fixtures__/boxedExpression.ts [hunk 11]
```diff
@@ -93,6 +98,36 @@ export class BoxedExpressionEditor {
     await this.page.getByRole("menuitem", { name: "Function" }).click();
   }
 
+  public async selectBoxedFilter(from: Page | Locator = this.page) {
+    this.select(from);
+    await this.page.getByRole("menuitem", { name: "Filter" }).click();
+  }
+
+  public async copyFilter(from: Page | Locator = this.page) {
+    await this.page.getByTestId("logic-type-button-test-id").click();
```
Comment (ljmotta) [inline packages/boxed-expression-component/tests-e2e/__fixtures__/boxedExpression.ts:110]: I guess @danielzhe will change this with the new API?

File: packages/boxed-expression-component/tests-e2e/features/selection/selection.spec.ts [hunk 12]
```diff
@@ -127,4 +127,31 @@ test.describe("Selection", () => {
       await expect(page.getByRole("cell", { name: "USA" })).toBeAttached();
     });
   });
+
+  test.describe.only("Filter Expression", () => {
```
Comment (ljmotta) [inline packages/boxed-expression-component/tests-e2e/features/selection/selection.spec.ts]: Need to remove the `only`.

**Review by ljmotta — COMMENTED** [review]
@jomarko Thanks for addressing the previous comments. I've made a few more inline. Also, looking at the screenshots, I don't see a case where we resize the the filter expression to accommodate the inner text. Resizing the `in` and the `match` would be ideal.

Just a heads up, I've unresolved a previous comment regarding the default `Literal Expression`.

File: packages/boxed-expression-component/stories/boxedExpressions/Filter/Filter.stories.tsx (hunk 3 above)
Comment (jomarko) [reply packages/boxed-expression-component/stories/boxedExpressions/Filter/Filter.stories.tsx]: ok, done

File: packages/boxed-expression-component/stories/getDefaultBoxedExpressionForStories.ts (hunk 9 above)
Comment (jomarko) [reply packages/boxed-expression-component/stories/getDefaultBoxedExpressionForStories.ts]: ok, done

File: packages/boxed-expression-component/tests-e2e/__screenshots__/Google-Chrome/features/resizing/boxed-filter-nested-resized-using-match.png
Comment (ljmotta) [inline packages/boxed-expression-component/tests-e2e/__screenshots__/Google-Chrome/features/resizing/boxed-filter-nested-resized-using-match.png:1]: @danielzhe It looks like a bug on the auto resize due to the `[` `]`? 👀

File: packages/boxed-expression-component/tests-e2e/boxedExpressions/filter/filterExpression.spec.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    await page.getByRole("option", { name: "boolean" }).click();
+    await page.keyboard.press("Enter");
+
+    await expect(page.getByRole("columnheader", { name: "Expression Name (boolean)" })).toBeVisible();
+  });
+
+  test("should correctly reset a nested filter", async ({ boxedExpressionEditor, page, stories }) => {
+    await stories.openBoxedFilter("nested");
+
+    await boxedExpressionEditor.resetFilter(page.locator("[data-ouia-component-id=expression-row-0]"));
```
Comment (ljmotta) [inline packages/boxed-expression-component/tests-e2e/boxedExpressions/filter/filterExpression.spec.ts]: We shouldn't be using the `data-ouia-component-id`.

**Review by ljmotta — CHANGES_REQUESTED** [review]
@jomarko Thanks for addressing all changes promptly. I've just saw one thing that slipped through my last review.

File: packages/boxed-expression-component/tests-e2e/boxedExpressions/filter/filterExpression.spec.ts (hunk 13 above)
Comment (jomarko) [reply packages/boxed-expression-component/tests-e2e/boxedExpressions/filter/filterExpression.spec.ts]: thank you for the feedback, changes pushed

**Review by ljmotta — APPROVED** [review]

**Review by tiagobento — APPROVED** [review]

**Pull request merged**